src/utils/debounce.ts

```diff
@@ -1,146 +1,29 @@
-interface DebounceOptions {
-  maxWait?: number;
-  leading?: boolean;
-  trailing?: boolean;
-}
-
-export interface IDebounceReturn {
-  cancel: () => void;
-  flush: () => void;
-  isPending: () => boolean;
-}
-
-export interface DebouncedReturn<T extends (...args: any[]) => ReturnType<T>> extends IDebounceReturn {
-  (...args: Parameters<T>): ReturnType<T> | undefined;
-}
-
-export function debounce<T extends (...args: any) => ReturnType<T>>(fn: T, wait?: number, options?: DebounceOptions): DebouncedReturn<T> {
-
-  let lastArgs;
-  let lastThis: unknown;
-  let timerId: number | null = null;
-  let lastCallTime: number | null = null;
-  let lastInvokeTime = 0;
-  let leading = false;
-  let maxing = false;
-  let trailing = true;
-  let result;
-
-  const useRAF = !wait && wait !== 0 && typeof window !== 'undefined' && typeof window.requestAnimationFrame === 'function';
-
-  if (typeof fn !== 'function') {
-    throw new TypeError('Expected a function');
-  }
-
-  wait = +wait! || 0;
-  options = options || {};
-
-  leading = !!options.leading;
-  trailing = 'trailing' in options ? !!options.trailing : true;
-  maxing = 'maxWait' in options;
-  const maxWait = maxing ? Math.max(+options.maxWait! || 0, wait) : null;
-
-  function invokeFunc(time: number) {
-    const args = lastArgs;
-    const thisArg = lastThis;
-
-    lastArgs = lastThis = null;
-    lastInvokeTime = time;
-    result = fn.apply(thisArg, args);
-    return result;
-  }
-
-  function startTimer(pendingFunc: () => void, wait: number) {
-    if (useRAF) {
-      window.cancelAnimationFrame(timerId!);
-      return window.requestAnimationFrame(pendingFunc);
-    }
-    return setTimeout(pendingFunc, wait);
-  }
 
-  function cancelTimer(id) {
-    useRAF ? window.cancelAnimationFrame(id) : clearTimeout(id);
-  }
+export function debounce(fn: Function, wait?: number, leading?: boolean) {
+  let timer;
+  let lastCalledAt = -Infinity;
```

Comment:
Why not use `null` or `undefined`? 🤔

Reply:
1. it has the type `number`,
2. you can perform arithmetic operations when assigning it to `Infinity`.